libenv: simplify and improve the removal of time-based classes

tzz opened the pull request:
This greatly simplifies the removal of time-based classes and makes it much faster and more reliable by using tags instead of blindly shooting down class names.

kacf commented:
Indeed much better. I don't even understand how the old function was supposed to work. You would just remove the same classes you were adding...

kacf commented:
trigger build

kacf commented:
Hmm, this seems to break the `./promises/autorun-D.cf` and `./promises/autorun-def_json.cf` tests in masterfiles pretty consistently. Master is in kind of a bad state right now though, so I can't guarantee 100% it's related. I'll be working to fix it, but perhaps you could check in the meantime if you also get the same failure.

kacf commented:
Yes, this is definitely failing.

tzz commented:
I can't even run the masterfiles tests anymore. Whatever autoconf/automake magic has been done there, on a recent Ubuntu system they're totally busted. I checked the core acceptance tests:

```
======================================================================
Testsuite finished at 2016-04-25 13:12:35 (756 seconds)

Passed tests:  1321
Failed tests:  0
Skipped tests: 263
Soft failures: 25
Total tests:   1609
```

Let me know when you know for sure that this code breaks something in masterfiles.

kacf commented:
Hmm, can you be more elaborate? Recently we switched to autodetecting the version in a branch, which requires Git to be present, but I don't think this should affect you. Otherwise I'm not aware of any problems with it. Maybe try rebasing all branches (including core) to make sure it's all up to date?

kacf commented:
And try a complete `git clean -dxf` and `./autogen.sh` as well.

tzz commented:
Yeah, OK, that fixed the make. So, running `make check` proceeds until:

```
   error: can not find package wrapper in provided location '/home/tzz/.cfagent/modules/packages/apt_get' or access to file is restricted: 2
   error: Can not set up wrapper for module: apt_get
```

I don't think that is an issue with this pull request, and looking at all the tests you said are failing, it's **extremely** unlikely they are related to this pull request because it only deals with time classes. Can you try again on your side?

kacf commented:
This is the error I get in `masterfiles/tests/acceptance/promises/autorun-D.cf`:

```
======================================================================
Testsuite started at 2016-05-02 08:49:08
----------------------------------------------------------------------
Total tests: 1

CRASHING_TESTS: enabled
NETWORK_TESTS: enabled
STAGING_TESTS: disabled
UNSAFE_TESTS: disabled
LIBXML2_TESTS: enabled

----------------------------------------------------------------------
./promises/autorun-D.cf
----------------------------------------------------------------------
  notice: Q: "...r/bin/make inst": Making install in modules/packages
Q: "...r/bin/make inst": make[1]: Entering directory `/root/code/masterfiles/modules/packages'
Q: "...r/bin/make inst": make[2]: Entering directory `/root/code/masterfiles/modules/packages'
Q: "...r/bin/make inst": make[2]: Nothing to be done for `install-exec-am'.
Q: "...r/bin/make inst": test -z "/var/cfengine/modules/packages" || /bin/mkdir -p "/root/code/masterfiles/tests/acceptance/workdir/__promises_autorun-D_cf/tmp/TESTDIR.cfengine/var/cfengine/modules/packages"
Q: "...r/bin/make inst":  /usr/bin/install -c apt_get yum pkgsrc freebsd_ports '/root/code/masterfiles/tests/acceptance/workdir/__promises_autorun-D_cf/tmp/TESTDIR.cfengine/var/cfengine/modules/packages'
Q: "...r/bin/make inst": make[2]: Leaving directory `/root/code/masterfiles/modules/packages'
Q: "...r/bin/make inst": make[1]: Leaving directory `/root/code/masterfiles/modules/packages'
Q: "...r/bin/make inst": Making install in tests/unit
Q: "...r/bin/make inst": make[1]: Entering directory `/root/code/masterfiles/tests/unit'
Q: "...r/bin/make inst": make[2]: Entering directory `/root/code/masterfiles/tests/unit'
Q: "...r/bin/make inst": make[2]: Nothing to be done for `install-exec-am'.
Q: "...r/bin/make inst": make[2]: Nothing to be done for `install-data-am'.
Q: "...r/bin/make inst": make[2]: Leaving directory `/root/code/masterfiles/tests/unit'
Q: "...r/bin/make inst": make[1]: Leaving directory `/root/code/masterfiles/tests/unit'
Q: "...r/bin/make inst": Making install in tests/acceptance
Q: "...r/bin/make inst": make[1]: Entering directory `/root/code/masterfiles/tests/acceptance'
Q: "...r/bin/make inst": make[2]: Entering directory `/root/code/masterfiles/tests/acceptance'
Q: "...r/bin/make inst": make[2]: Nothing to be done for `install-exec-am'.
Q: "...r/bin/make inst": make[2]: Nothing to be done for `install-data-am'.
Q: "...r/bin/make inst": make[2]: Leaving directory `/root/code/masterfiles/tests/acceptance'
Q: "...r/bin/make inst": make[1]: Leaving directory `/root/code/masterfiles/tests/acceptance'
Q: "...r/bin/make inst": make[1]: Entering directory `/root/code/masterfiles'
Q: "...r/bin/make inst": make[2]: Entering directory `/root/code/masterfiles'
Q: "...r/bin/make inst": make[2]: Nothing to be done for `install-exec-am'.
Q: "...r/bin/make inst": test -z "/var/cfengine/masterfiles" || /bin/mkdir -p "/root/code/masterfiles/tests/acceptance/workdir/__promises_autorun-D_cf/tmp/TESTDIR.cfengine/var/cfengine/masterfiles"
Q: "...r/bin/make inst": /bin/mkdir -p '/root/code/masterfiles/tests/acceptance/workdir/__promises_autorun-D_cf/tmp/TESTDIR.cfengine/var/cfengine/masterfiles/cfe_internal/enterprise/templates'
Q: "...r/bin/make inst":  /usr/bin/install -c -m 644  ./cfe_internal/enterprise/templates/httpd.conf.mustache '/root/code/masterfiles/tests/acceptance/workdir/__promises_autorun-D_cf/tmp/TESTDIR.cfengine/var/cfengine/masterfiles/cfe_internal/enterprise/templates'
Q: "...r/bin/make inst": /bin/mkdir -p '/root/code/masterfiles/tests/acceptance/workdir/__promises_autorun-D_cf/tmp/TESTDIR.cfengine/var/cfengine/masterfiles/cfe_internal/core/watchdog'
Q: "...r/bin/make inst":  /usr/bin/install -c -m 644  ./cfe_internal/core/watchdog/watchdog.cf '/root/code/masterfiles/tests/acceptance/workdir/__promises_autorun-D_cf/tmp/TESTDIR.cfengine/var/cfengine/masterfiles/cfe_internal/core/watchdog'
Q: "...r/bin/make inst": /bin/mkdir -p '/root/code/masterfiles/tests/acceptance/workdir/__promises_autorun-D_cf/tmp/TESTDIR.cfengine/var/cfengine/masterfiles/cfe_internal/core'
Q: "...r/bin/make inst":  /usr/bin/install -c -m 644  ./cfe_internal/core/log_rotation.cf ./cfe_internal/core/limit_robot_agents.cf ./cfe_internal/core/host_info_report.cf ./cfe_internal/core/main.cf '/root/code/masterfiles/tests/acceptance/workdir/__promises_autorun-D_cf/tmp/TESTDIR.cfengine/var/cfengine/masterfiles/cfe_internal/core'
Q: "...r/bin/make inst": /bin/mkdir -p '/root/code/masterfiles/tests/acceptance/workdir/__promises_autorun-D_cf/tmp/TESTDIR.cfengine/var/cfengine/masterfiles/cfe_internal/core/deprecated'
  notice: Q: "...r/bin/make inst":  /usr/bin/install -c -m 644  ./cfe_internal/core/deprecated/cfengine_processes.cf '/root/code/masterfiles/tests/acceptance/workdir/__promises_autorun-D_cf/tmp/TESTDIR.cfengine/var/cfengine/masterfiles/cfe_internal/core/deprecated'
Q: "...r/bin/make inst": /bin/mkdir -p '/root/code/masterfiles/tests/acceptance/workdir/__promises_autorun-D_cf/tmp/TESTDIR.cfengine/var/cfengine/masterfiles/templates'
Q: "...r/bin/make inst":  /usr/bin/install -c -m 644  ./templates/host_info_report.mustache ./templates/cfengine_watchdog.mustache '/root/code/masterfiles/tests/acceptance/workdir/__promises_autorun-D_cf/tmp/TESTDIR.cfengine/var/cfengine/masterfiles/templates'
Q: "...r/bin/make inst": /bin/mkdir -p '/root/code/masterfiles/tests/acceptance/workdir/__promises_autorun-D_cf/tmp/TESTDIR.cfengine/var/cfengine/masterfiles/cfe_internal'
Q: "...r/bin/make inst":  /usr/bin/install -c -m 644  ./cfe_internal/CFE_cfengine.cf '/root/code/masterfiles/tests/acceptance/workdir/__promises_autorun-D_cf/tmp/TESTDIR.cfengine/var/cfengine/masterfiles/cfe_internal'
Q: "...r/bin/make inst": /bin/mkdir -p '/root/code/masterfiles/tests/acceptance/workdir/__promises_autorun-D_cf/tmp/TESTDIR.cfengine/var/cfengine/masterfiles/controls/3.6'
Q: "...r/bin/make inst":  /usr/bin/install -c -m 644  ./controls/3.6/cf_serverd.cf ./controls/3.6/cf_hub.cf ./controls/3.6/cf_runagent.cf ./controls/3.6/reports.cf ./controls/3.6/cf_agent.cf ./controls/3.6/cf_execd.cf ./controls/3.6/def.cf ./controls/3.6/update_def_inputs.cf ./controls/3.6/cf_monitord.cf ./controls/3.6/def_inputs.cf ./controls/3.6/update_def.cf '/root/code/masterfiles/tests/acceptance/workdir/__promises_autorun-D_cf/tmp/TESTDIR.cfengine/var/cfengine/masterfiles/controls/3.6'
Q: "...r/bin/make inst": /bin/mkdir -p '/root/code/masterfiles/tests/acceptance/workdir/__promises_autorun-D_cf/tmp/TESTDIR.cfengine/var/cfengine/masterfiles/cfe_internal/update'
Q: "...r/bin/make inst":  /usr/bin/install -c -m 644  ./cfe_internal/update/update_bins.cf ./cfe_internal/update/cfe_internal_dc_workflow.cf ./cfe_internal/update/cfe_internal_local_git_remote.cf ./cfe_internal/update/update_processes.cf ./cfe_internal/update/cfe_internal_update_from_repository.cf ./cfe_internal/update/update_policy.cf '/root/code/masterfiles/tests/acceptance/workdir/__promises_autorun-D_cf/tmp/TESTDIR.cfengine/var/cfengine/masterfiles/cfe_internal/update'
Q: "...r/bin/make inst": /bin/mkdir -p '/root/code/masterfiles/tests/acceptance/workdir/__promises_autorun-D_cf/tmp/TESTDIR.cfengine/var/cfengine/masterfiles/cfe_internal/enterprise/ha'
Q: "...r/bin/make inst":  /usr/bin/install -c -m 644  ./cfe_internal/enterprise/ha/ha.cf ./cfe_internal/enterprise/ha/ha_update.cf ./cfe_internal/enterprise/ha/ha_def.cf '/root/code/masterfiles/tests/acceptance/workdir/__promises_autorun-D_cf/tmp/TESTDIR.cfengine/var/cfengine/masterfiles/cfe_internal/enterprise/ha'
Q: "...r/bin/make inst": /bin/mkdir -p '/root/code/masterfiles/tests/acceptance/workdir/__promises_autorun-D_cf/tmp/TESTDIR.cfengine/var/cfengine/masterfiles/lib/3.6'
Q: "...r/bin/make inst":  /usr/bin/install -c -m 644  ./lib/3.6/common.cf ./lib/3.6/cfe_internal.cf ./lib/3.6/users.cf ./lib/3.6/processes.cf ./lib/3.6/paths.cf ./lib/3.6/monitor.cf ./lib/3.6/cfe_internal_hub.cf ./lib/3.6/cfengine_enterprise_hub_ha.cf ./lib/3.6/files.cf ./lib/3.6/packages.cf ./lib/3.6/stdlib.cf ./lib/3.6/commands.cf ./lib/3.6/bundles.cf ./lib/3.6/feature.cf ./lib/3.6/guest_environments.cf ./lib/3.6/services.cf ./lib/3.6/autorun.cf ./lib/3.6/examples.cf ./lib/3.6/edit_xml.cf ./lib/3.6/vcs.cf ./lib/3.6/storage.cf ./lib/3.6/databases.cf '/root/code/masterfiles/tests/acceptance/workdir/__promises_autorun-D_cf/tmp/TESTDIR.cfengine/var/cfengine/masterfiles/lib/3.6'
Q: "...r/bin/make inst": /bin/mkdir -p '/root/code/masterfiles/tests/acceptance/workdir/__promises_autorun-D_cf/tmp/TESTDIR.cfengine/var/cfengine/masterfiles/lib/3.7'
  notice: Q: "...r/bin/make inst":  /usr/bin/install -c -m 644  lib/3.7/README.md '/root/code/masterfiles/tests/acceptance/workdir/__promises_autorun-D_cf/tmp/TESTDIR.cfengine/var/cfengine/masterfiles/lib/3.7'
Q: "...r/bin/make inst":  /usr/bin/install -c -m 644  ./update.cf ./promises.cf '/root/code/masterfiles/tests/acceptance/workdir/__promises_autorun-D_cf/tmp/TESTDIR.cfengine/var/cfengine/masterfiles/.'
Q: "...r/bin/make inst": /bin/mkdir -p '/root/code/masterfiles/tests/acceptance/workdir/__promises_autorun-D_cf/tmp/TESTDIR.cfengine/var/cfengine/masterfiles/controls'
Q: "...r/bin/make inst":  /usr/bin/install -c -m 644  ./controls/cf_serverd.cf ./controls/cf_hub.cf ./controls/cf_runagent.cf ./controls/reports.cf ./controls/cf_agent.cf ./controls/cf_execd.cf ./controls/def.cf ./controls/update_def_inputs.cf ./controls/cf_monitord.cf ./controls/def_inputs.cf ./controls/update_def.cf '/root/code/masterfiles/tests/acceptance/workdir/__promises_autorun-D_cf/tmp/TESTDIR.cfengine/var/cfengine/masterfiles/controls'
Q: "...r/bin/make inst": /bin/mkdir -p '/root/code/masterfiles/tests/acceptance/workdir/__promises_autorun-D_cf/tmp/TESTDIR.cfengine/var/cfengine/masterfiles/services/autorun'
Q: "...r/bin/make inst":  /usr/bin/install -c -m 644  ./services/autorun/hello.cf '/root/code/masterfiles/tests/acceptance/workdir/__promises_autorun-D_cf/tmp/TESTDIR.cfengine/var/cfengine/masterfiles/services/autorun'
Q: "...r/bin/make inst": /bin/mkdir -p '/root/code/masterfiles/tests/acceptance/workdir/__promises_autorun-D_cf/tmp/TESTDIR.cfengine/var/cfengine/masterfiles/cfe_internal/enterprise'
Q: "...r/bin/make inst":  /usr/bin/install -c -m 644  ./cfe_internal/enterprise/CFE_hub_specific.cf ./cfe_internal/enterprise/mission_portal.cf ./cfe_internal/enterprise/CFE_knowledge.cf ./cfe_internal/enterprise/file_change.cf ./cfe_internal/enterprise/main.cf '/root/code/masterfiles/tests/acceptance/workdir/__promises_autorun-D_cf/tmp/TESTDIR.cfengine/var/cfengine/masterfiles/cfe_internal/enterprise'
Q: "...r/bin/make inst": /bin/mkdir -p '/root/code/masterfiles/tests/acceptance/workdir/__promises_autorun-D_cf/tmp/TESTDIR.cfengine/var/cfengine/masterfiles/sketches/meta'
Q: "...r/bin/make inst":  /usr/bin/install -c -m 644  ./sketches/meta/api-runfile.cf '/root/code/masterfiles/tests/acceptance/workdir/__promises_autorun-D_cf/tmp/TESTDIR.cfengine/var/cfengine/masterfiles/sketches/meta'
Q: "...r/bin/make inst": /bin/mkdir -p '/root/code/masterfiles/tests/acceptance/workdir/__promises_autorun-D_cf/tmp/TESTDIR.cfengine/var/cfengine/masterfiles/services'
Q: "...r/bin/make inst":  /usr/bin/install -c -m 644  ./services/main.cf '/root/code/masterfiles/tests/acceptance/workdir/__promises_autorun-D_cf/tmp/TESTDIR.cfengine/var/cfengine/masterfiles/services'
Q: "...r/bin/make inst": /bin/mkdir -p '/root/code/masterfiles/tests/acceptance/workdir/__promises_autorun-D_cf/tmp/TESTDIR.cfengine/var/cfengine/masterfiles/inventory'
Q: "...r/bin/make inst":  /usr/bin/install -c -m 644  ./inventory/linux.cf ./inventory/macos.cf ./inventory/os.cf ./inventory/suse.cf ./inventory/generic.cf ./inventory/lsb.cf ./inventory/redhat.cf ./inventory/freebsd.cf ./inventory/windows.cf ./inventory/debian.cf ./inventory/any.cf '/root/code/masterfiles/tests/acceptance/workdir/__promises_autorun-D_cf/tmp/TESTDIR.cfengine/var/cfengine/masterfiles/inventory'
Q: "...r/bin/make inst": /bin/mkdir -p '/root/code/masterfiles/tests/acceptance/workdir/__promises_autorun-D_cf/tmp/TESTDIR.cfengine/var/cfengine/masterfiles/lib'
  notice: Q: "...r/bin/make inst":  /usr/bin/install -c -m 644  ./lib/common.cf ./lib/cfe_internal.cf ./lib/users.cf ./lib/processes.cf ./lib/paths.cf ./lib/monitor.cf ./lib/cfe_internal_hub.cf ./lib/cfengine_enterprise_hub_ha.cf ./lib/files.cf ./lib/packages.cf ./lib/stdlib.cf ./lib/reports.cf ./lib/commands.cf ./lib/bundles.cf ./lib/feature.cf ./lib/guest_environments.cf ./lib/services.cf ./lib/autorun.cf ./lib/examples.cf ./lib/edit_xml.cf ./lib/vcs.cf ./lib/storage.cf ./lib/databases.cf '/root/code/masterfiles/tests/acceptance/workdir/__promises_autorun-D_cf/tmp/TESTDIR.cfengine/var/cfengine/masterfiles/lib'
Q: "...r/bin/make inst": make[2]: Leaving directory `/root/code/masterfiles'
Q: "...r/bin/make inst": make[1]: Leaving directory `/root/code/masterfiles'
R: /root/code/masterfiles/tests/acceptance/./promises/autorun-D.cf FAIL
R: dcs_passif: failing based on class '__passif'
R: dcs_passif_expected: bad: __passif_output_wanted was NOT defined
R: dcs_passif_expected: bad: __passif_output_unwanted WAS defined
R: dcs_passif_output: bad: '.*R: hello_world_autorun: Hello, this is an automatically loaded bundle.*' was NOT found in the output of '"/root/code/masterfiles/tests/acceptance/workdir/__promises_autorun-D_cf/bin/cf-agent" -Dservices_autorun -v -f /root/code/masterfiles/tests/acceptance/workdir/__promises_autorun-D_cf/tmp/TESTDIR.cfengine/var/cfengine/masterfiles/promises.cf | /bin/grep Hello'
R: dcs_passif_output: bad: '' WAS found in the output of '"/root/code/masterfiles/tests/acceptance/workdir/__promises_autorun-D_cf/bin/cf-agent" -Dservices_autorun -v -f /root/code/masterfiles/tests/acceptance/workdir/__promises_autorun-D_cf/tmp/TESTDIR.cfengine/var/cfengine/masterfiles/promises.cf | /bin/grep Hello'
R: dcs_passif_output: the output of command '"/root/code/masterfiles/tests/acceptance/workdir/__promises_autorun-D_cf/bin/cf-agent" -Dservices_autorun -v -f /root/code/masterfiles/tests/acceptance/workdir/__promises_autorun-D_cf/tmp/TESTDIR.cfengine/var/cfengine/masterfiles/promises.cf | /bin/grep Hello' was: ''

Return code is 0.

  ==> FAIL (UNEXPECTED FAILURE)


======================================================================
Testsuite finished at 2016-05-02 08:49:12 (4 seconds)

Passed tests:  0
Failed tests:  1
Skipped tests: 0
Soft failures: 0
Total tests:   1
```

Without your branch, it passes.

tzz commented:
OK; fixed.

The bug was that when we search for classes, **all** the tags are OR-ed. So I was removing all classes with `source=agent` originally because I thought the tags were AND-ed.

Rebased and ready for another review.

kacf commented:
Oh, hehe, nice catch!

kacf commented:
trigger build

kacf commented:
Thanks, all green!
